Race at tip

@gernest Please try if you can reproduce the race.

Using https://github.com/cznic/ql/commit/26248eaae4fd0cdad403986b3c77eff575b20945

```bash
$ go test -race -timeout 0
==================
WARNING: DATA RACE
Write at 0x00c420016870 by goroutine 22:
  github.com/cznic/ql.(*mem).Close()
      /home/jnml/src/github.com/cznic/ql/mem.go:284 +0xa2
  github.com/cznic/ql.(*DB).Close()
      /home/jnml/src/github.com/cznic/ql/ql.go:1442 +0xf7
  github.com/cznic/ql.(*driverConn).Close()
      /home/jnml/src/github.com/cznic/ql/driver.go:267 +0x3f8
  database/sql.(*driverConn).finalClose.func2()
      /home/jnml/go/src/database/sql/sql.go:440 +0x8c
  database/sql.withLock()
      /home/jnml/go/src/database/sql/sql.go:2545 +0x76
  database/sql.(*driverConn).finalClose()
      /home/jnml/go/src/database/sql/sql.go:442 +0x215
  database/sql.(finalCloser).(database/sql.finalClose)-fm()
      /home/jnml/go/src/database/sql/sql.go:537 +0x46
  database/sql.(*DB).Close()
      /home/jnml/go/src/database/sql/sql.go:644 +0x4a8
  github.com/cznic/ql.TestWhereExists()
      /home/jnml/src/github.com/cznic/ql/all_test.go:3504 +0x4b9
  testing.tRunner()
      /home/jnml/go/src/testing/testing.go:657 +0x107

Previous read at 0x00c420016870 by goroutine 28:
  github.com/cznic/ql.(*mem).Read()
      /home/jnml/src/github.com/cznic/ql/mem.go:456 +0x64
  github.com/cznic/ql.(*table).row0()
      /home/jnml/src/github.com/cznic/ql/storage.go:769 +0xe6
  github.com/cznic/ql.(*tableDefaultPlan).do()
      /home/jnml/src/github.com/cznic/ql/plan.go:2403 +0xcb
  github.com/cznic/ql.(*DB).do()
      /home/jnml/src/github.com/cznic/ql/ql.go:1474 +0x1f5
  github.com/cznic/ql.recordset.Do()
      /home/jnml/src/github.com/cznic/ql/ql.go:75 +0x96
  github.com/cznic/ql.(*recordset).Do()
      <autogenerated>:60 +0xa3
  github.com/cznic/ql.newdriverRows.func1()
      /home/jnml/src/github.com/cznic/ql/driver.go:422 +0x106

Goroutine 22 (running) created at:
  testing.(*T).Run()
      /home/jnml/go/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /home/jnml/go/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /home/jnml/go/src/testing/testing.go:657 +0x107
  testing.runTests()
      /home/jnml/go/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /home/jnml/go/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/cznic/ql/_test/_testmain.go:346 +0x20f

Goroutine 28 (finished) created at:
  github.com/cznic/ql.newdriverRows()
      /home/jnml/src/github.com/cznic/ql/driver.go:430 +0x13d
  github.com/cznic/ql.newDriverMultiRows()
      /home/jnml/src/github.com/cznic/ql/driver.go:524 +0x67
  github.com/cznic/ql.driverQuery()
      /home/jnml/src/github.com/cznic/ql/driver.go:379 +0x1e0
  github.com/cznic/ql.(*driverStmt).Query()
      /home/jnml/src/github.com/cznic/ql/driver.go:592 +0x10c
  database/sql.ctxDriverStmtQuery()
      /home/jnml/go/src/database/sql/ctxutil.go:95 +0x1e9
  database/sql.rowsiFromStatement()
      /home/jnml/go/src/database/sql/sql.go:2002 +0x283
  database/sql.(*Stmt).QueryContext()
      /home/jnml/go/src/database/sql/sql.go:1949 +0x2a3
  database/sql.(*Stmt).QueryRowContext()
      /home/jnml/go/src/database/sql/sql.go:2021 +0x80
  database/sql.(*Stmt).QueryRow()
      /home/jnml/go/src/database/sql/sql.go:2040 +0x82
  github.com/cznic/ql.TestWhereExists.func1()
      /home/jnml/src/github.com/cznic/ql/all_test.go:3486 +0x11f
==================
==================
WARNING: DATA RACE
Write at 0x00c420018730 by goroutine 22:
  github.com/cznic/ql.(*DB).Close()
      /home/jnml/src/github.com/cznic/ql/ql.go:1443 +0x14a
  github.com/cznic/ql.(*driverConn).Close()
      /home/jnml/src/github.com/cznic/ql/driver.go:267 +0x3f8
  database/sql.(*driverConn).finalClose.func2()
      /home/jnml/go/src/database/sql/sql.go:440 +0x8c
  database/sql.withLock()
      /home/jnml/go/src/database/sql/sql.go:2545 +0x76
  database/sql.(*driverConn).finalClose()
      /home/jnml/go/src/database/sql/sql.go:442 +0x215
  database/sql.(finalCloser).(database/sql.finalClose)-fm()
      /home/jnml/go/src/database/sql/sql.go:537 +0x46
  database/sql.(*DB).Close()
      /home/jnml/go/src/database/sql/sql.go:644 +0x4a8
  github.com/cznic/ql.TestWhereExists()
      /home/jnml/src/github.com/cznic/ql/all_test.go:3504 +0x4b9
  testing.tRunner()
      /home/jnml/go/src/testing/testing.go:657 +0x107

Previous read at 0x00c420018730 by goroutine 28:
  github.com/cznic/ql.(*table).row0()
      /home/jnml/src/github.com/cznic/ql/storage.go:769 +0x6b
  github.com/cznic/ql.(*tableDefaultPlan).do()
      /home/jnml/src/github.com/cznic/ql/plan.go:2403 +0xcb
  github.com/cznic/ql.(*DB).do()
      /home/jnml/src/github.com/cznic/ql/ql.go:1474 +0x1f5
  github.com/cznic/ql.recordset.Do()
      /home/jnml/src/github.com/cznic/ql/ql.go:75 +0x96
  github.com/cznic/ql.(*recordset).Do()
      <autogenerated>:60 +0xa3
  github.com/cznic/ql.newdriverRows.func1()
      /home/jnml/src/github.com/cznic/ql/driver.go:422 +0x106

Goroutine 22 (running) created at:
  testing.(*T).Run()
      /home/jnml/go/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /home/jnml/go/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /home/jnml/go/src/testing/testing.go:657 +0x107
  testing.runTests()
      /home/jnml/go/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /home/jnml/go/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/cznic/ql/_test/_testmain.go:346 +0x20f

Goroutine 28 (finished) created at:
  github.com/cznic/ql.newdriverRows()
      /home/jnml/src/github.com/cznic/ql/driver.go:430 +0x13d
  github.com/cznic/ql.newDriverMultiRows()
      /home/jnml/src/github.com/cznic/ql/driver.go:524 +0x67
  github.com/cznic/ql.driverQuery()
      /home/jnml/src/github.com/cznic/ql/driver.go:379 +0x1e0
  github.com/cznic/ql.(*driverStmt).Query()
      /home/jnml/src/github.com/cznic/ql/driver.go:592 +0x10c
  database/sql.ctxDriverStmtQuery()
      /home/jnml/go/src/database/sql/ctxutil.go:95 +0x1e9
  database/sql.rowsiFromStatement()
      /home/jnml/go/src/database/sql/sql.go:2002 +0x283
  database/sql.(*Stmt).QueryContext()
      /home/jnml/go/src/database/sql/sql.go:1949 +0x2a3
  database/sql.(*Stmt).QueryRowContext()
      /home/jnml/go/src/database/sql/sql.go:2021 +0x80
  database/sql.(*Stmt).QueryRow()
      /home/jnml/go/src/database/sql/sql.go:2040 +0x82
  github.com/cznic/ql.TestWhereExists.func1()
      /home/jnml/src/github.com/cznic/ql/all_test.go:3486 +0x11f
==================
--- FAIL: TestWhereExists (0.01s)
	testing.go:610: race detected during execution of test
FAIL
exit status 1
FAIL	github.com/cznic/ql	89.349s
$ 
```

--- a/mem.go
+++ b/mem.go
@@ -11,6 +11,7 @@ import (
 	"fmt"
 	"io"
 	"math/big"
+	"sync"
 	"time"
 )
 
@@ -247,6 +248,7 @@ type mem struct {
 	recycler []int
 	tnl      int
 	rollback *undos
+	mu       sync.RWMutex
 }
 
 func newMemStorage() (s *mem, err error) {
@@ -278,10 +280,13 @@ func (s *mem) newUndo(tag int, h int64, data []interface{}) {
 func (s *mem) Acid() bool { return false }
 
 func (s *mem) Close() (err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	if s.tnl != 0 {
 		return fmt.Errorf("cannot close DB while open transaction exist")
 	}
-	*s = mem{}
+	s.data, s.recycler, s.rollback = nil, nil, nil
+	s.id, s.tnl = 0, 0
 	return
 }
 
@@ -453,6 +458,8 @@ func (s *mem) Create(data ...interface{}) (h int64, err error) {
 }
 
 func (s *mem) Read(dst []interface{}, h int64, cols ...*col) (data []interface{}, err error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 	if i := int(h); i != 0 && i < len(s.data) {
 		d := s.clone(s.data[h]...)
 		if cols == nil {

